### PR TITLE
documentation #84: updated swagger for teachers

### LINKED ISSUE #84: [DOC] Write Swagger doc for Teacher
We are updating the documentation for all endpoints and adding more _description and details_. 

One of the main controllers is Teacher. You should update the documentation according to the **pattern** below:
```ts
  @ApiBearerAuth()
  @ApiOkResponse({
    type: DisciplineTeachersResponse,
  })
  @ApiBadRequestResponse({
    description: `\n
    InvalidQueryException:
      Type of discipline must be a field of enum

    InvalidDisciplineIdException:
      Discipline with such id is not found`,
  })
  @ApiUnauthorizedResponse({
    description: `\n
    UnauthorizedException:
      Unauthorized`,
  })
  @ApiForbiddenResponse({
    description: `\n
    NoPermissionException:
      You do not have permission to perform this action`,
  })
  @ApiParam({
    name: 'disciplineId',
    required: true,
    description: 'Id of certain discipline',
  })
  @ApiQuery({
    name: 'disciplineType',
    required: true,
    enum: DisciplineTypeEnum,
    description: 'Discipline type of some discipline',
  })
 @ApiEndpoint({
    summary: 'Receive teachers by discipline',
    permissions: PERMISSION.GROUPS_$GROUPID_DISCIPLINES_TEACHERS_GET,
  })
  @Get('/:disciplineId/teachers')
  async getAllByDiscipline (
    @Param('disciplineId', DisciplineByIdPipe) disciplineId: string,
    @Query('disciplineType') disciplineType: DisciplineTypeEnum
  ) {
    const teachers = await this.disciplineService.getTeachers(disciplineId, disciplineType);
    return { teachers };
  }
```
**Structure:**
- auth (if needed)
- ok response
- client errors
- params/query description
- access (if needed)

You should also add **field description** in the bady or response.
Add **summary**  to endpoints
Set **default values** for the optional fields

image

_Buena suerte_

[Our Swagger](https://apidev.fictadvisor.com/api)
[Doc](https://docs.nestjs.com/openapi/introduction)

### PR DESCRIPTION
**Pull request by VlO-OlV**
- updated getTeacherRoles endpoint
- updated getSubjects endpoint
- updated getDisciplines endpoint
- updated getSubject endpoint
- updated getTeacher endpoint
- updated response body in changeGroup endpoint

closes #84

### CODE REVIEW

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 1]
```diff
@@ -96,6 +96,14 @@ export class TeacherController {
     description: `InvalidEntityId:\n
```
Comment (dmykyr) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:96]: Incorrect response descritpion. It should look like: 
```
description: `\n
response_name:
  response_possible_text`,
```

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 2]
```diff
@@ -112,6 +120,14 @@ export class TeacherController {
...
```
Comment (dmykyr) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:120]: Incorrect response descritpion. It should look like: 
```
description: `\n
response_name:
  response_possible_text`,
```

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 3]
```diff
@@ -143,6 +150,24 @@ export class TeacherController {
     description: `NoPermissionException:\n
```
Comment (dmykyr) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:150]: Incorrect response descritpion. It should look like: 
```
description: `\n
response_name:
  response_possible_text`,
```

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 4]
```diff
@@ -161,6 +186,19 @@ export class TeacherController {
                   teacher with such id is not found
```
Comment (dmykyr) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:186]: Incorrect response descritpion. It should look like: 
```
description: `\n
response_name:
  response_possible_text`,
```

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 5]
```diff
@@ -176,6 +214,14 @@ export class TeacherController {
     description: `InvalidEntityIdException:\n
```
Comment (dmykyr) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:214]: Incorrect response descritpion. It should look like: 
```
description: `\n
response_name:
  response_possible_text`,
```

**Review by dmykyr — CHANGES_REQUESTED** [review]
Also change exception throwing in TeacherByIdPipe file.  It should look like: 
`throw new InvalidEntityIdException('teacher');`

**Review by dmykyr — CHANGES_REQUESTED** [review]
Additionaly you need to change exception throwing in SubjectByIdPipe file. It should look like: `throw new InvalidEntityIdException('Subject');`

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 6]
```diff
@@ -157,9 +185,22 @@ export class TeacherController {
     type: TeacherWithSubjectResponse,
   })
   @ApiBadRequestResponse({
-    description: `InvalidEntityIdException:\n
-                  teacher with such id is not found
-                  subject with such id is not found`,
+    description: `\n
+    InvalidEntityIdException:
+      Teacher with such id is not found subject with such id is not found`,
```
Comment (dmykyr) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:190]: All entities in InvalidEntityIdException must start with capital letter, so second text should look like: 
`Subject with such id is not found`

**Review by dmykyr — CHANGES_REQUESTED** [review]

**Comment by illiakroshka** [comment]
One more task you need to update response body in this endpoint:
image

**Review by dmykyr — APPROVED** [review]

**Review by illiakroshka — APPROVED** [review]

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 7]
```diff
@@ -93,8 +93,17 @@ export class TeacherController {
     type: TeacherRolesResponse,
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:93]: Pls add description
image

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 8]
```diff
@@ -109,8 +118,17 @@ export class TeacherController {
     type: SubjectsResponse,
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:118]: Also add description
image

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 9]
```diff
@@ -121,27 +139,38 @@ export class TeacherController {
...
+  })
+  @ApiParam({
+    name: 'teacherId',
+    required: true,
+    description: 'Id of certain teacher',
+  })
+  @ApiQuery({
+    name: 'notAnswered',
+    type: Boolean,
+  })
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:165]: Add description

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 10]
```diff
@@ -121,27 +139,38 @@ export class TeacherController {
...
-  @ApiQuery({
-    type: Boolean,
-    name: 'notAnswered',
-  })
-  @ApiQuery({
-    type: String,
-    name: 'userId',
-  })
   @ApiOkResponse({
     type: DisciplineTeacherAndSubjectResponse,
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:144]: Add descriptions
image

File: fictadvisor-api/src/v2/api/controllers/TeacherController.ts [hunk 11]
```diff
@@ -121,27 +139,38 @@ export class TeacherController {
...
+  @ApiQuery({
+    name: 'userId',
+    type: String,
+    description: 'Id of certain user',
+  })
+  @ApiEndpoint({
+    summary: 'Receive disciplines by teacher',
+    permissions: PERMISSION.USERS_$USERID_TEACHERS_$TEACHERID_DISCIPLINES_GET,
   })
   @Get('/:teacherId/disciplines')
```
Comment (fokaaas) [inline fictadvisor-api/src/v2/api/controllers/TeacherController.ts:175]: Also add unauthorized exception

**Review by fokaaas — CHANGES_REQUESTED** [review]

**Review by fokaaas — APPROVED** [review]

**Pull request merged**